### PR TITLE
Improvement on bitcoinj for managing many account number (arbitrary accout path)

### PR DESCRIPTION
**Pull request by gmagnotta**
Dear All,

I performed some changes on bitcoinj to implement the managing of many
account number (account path) not only the fixed
org.bitcoinj.wallet.DeterministicKeyChain.ACCOUNT_ZERO_PATH.

Please, can you evaluate it and include upstream or suggest me how to improve it in order to be
accepted?

Hope to receive feedback.

Thank you

Best regards,
Giuseppe

giuseppe.magnotta@gmail.com

### CODE REVIEW

**Comment by schildbach** [comment]
Thanks a lot, this PR is much appreciated! I'll have a look at it in the coming days.

**Comment by schildbach** [comment]
Just out of curiosity: What path are you using this code with? And maybe you would like to add one or two unit tests for these alternate path(s)?

**Comment by gmagnotta** [comment]
Hi,

I'm adding more nodes at the end of the normal path for additional keys associated at each address.

 Yes, I will add more unit tests.


Regards,
Giuseppe

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 1]
```diff
@@ -420,8 +445,19 @@ protected DeterministicKeyChain(KeyCrypter crypter, KeyParameter aesKey, Determi
 
     /** Override in subclasses to use a different account derivation path */
     protected ImmutableList<ChildNumber> getAccountPath() {
+    		if (accountPath != null) {
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java]: Can you use spaces for indentation here?

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 2]
```diff
@@ -420,8 +445,19 @@ protected DeterministicKeyChain(KeyCrypter crypter, KeyParameter aesKey, Determi
...
+    		}
+    		
         return ACCOUNT_ZERO_PATH;
     }
+    
+    /*
+     * Store account path of this DeterministicKey
+     */
+    protected void setAccountPath(ImmutableList<ChildNumber> accountPath) {
+    		this.accountPath = accountPath;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java]: Spaces again.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java [hunk 3]
```diff
@@ -292,6 +299,17 @@ public static Wallet fromWatchingKeyB58(NetworkParameters params, String watchKe
     }
 
     /**
+     * Creates a wallet that tracks payments to and from the HD key hierarchy rooted by the given watching key. The account path
+     * is specified. The key is specified in base58 notation and the creation time of the key. If you don't know the creation time,
+     * you can pass {@link DeterministicHierarchy#BIP32_STANDARDISATION_TIME_SECS}.
+     */
+    public static Wallet fromWatchingKeyB58(NetworkParameters params, String watchKeyB58, long creationTimeSeconds, ImmutableList<ChildNumber> accountPath) {
+    		final DeterministicKey watchKey = DeterministicKey.deserializeB58(null, watchKeyB58, params);
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/wallet/Wallet.java]: And spaces here, too.

**Comment by schildbach** [comment]
Generally looks good and unintrusive to me, which is a good thing! I just noticed your PR is based on the release branch, which will not work. New stuff is getting added to master first, so can you rebase your branch to master?

Do you still plan to add tests? I personally would like to test if your improvement is suited to manage one account of a BIP43+44 wallet and will perhaps write a test for that myself.

**Comment by gmagnotta** [comment]
Dear Andreas @schildbach ,

I performed the merge in the master branch. However, should I submit to you a new PR from master branch?
Or is it possible to link the new PR to this 'old'?

Thank you!

Regards,
Giuseppe

**Comment by schildbach** [comment]
I think you need to open a new PR for switching the base.

**Comment by schildbach** [comment]
Closed in favor of #1331.

**Pull request closed without merging**